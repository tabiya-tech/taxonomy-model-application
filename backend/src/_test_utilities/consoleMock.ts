/* eslint-disable @typescript-eslint/no-empty-function */
export {};
// Suppress chatty console during the tests
jest.spyOn(console, "error").mockImplementation(() => {});
jest.spyOn(console, "warn").mockImplementation(() => {});
jest.spyOn(console, "log").mockImplementation(() => {});
jest.spyOn(console, "info").mockImplementation(() => {});
