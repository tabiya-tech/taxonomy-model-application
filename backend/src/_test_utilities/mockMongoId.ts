import mongoose from "mongoose";

export function getMockStringId(index: number): string {
  return index.toString(16).padStart(24, "0");
}

export function getMockObjectId(index: number): mongoose.Types.ObjectId {
  return mongoose.Types.ObjectId.createFromHexString(getMockStringId(index));
}
