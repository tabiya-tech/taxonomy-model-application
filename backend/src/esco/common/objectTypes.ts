/**
 * Enum for the different types of objects in the ESCO ontology.
 */
export enum ObjectTypes {
  ISCOGroup = "ISCOGroup",
  Occupation = "Occupation",
  Skill = "Skill",
  SkillGroup = "SkillGroup",
}

/**
 * Enum for the two different types of relations in the ESCO ontology, essential and optional
 */
export enum RelationType {
  ESSENTIAL = "essential",
  OPTIONAL = "optional",
}

/**
 * Describes the three Types of Occupations
 */

export enum OccupationType {
  ESCO = "ESCO",
  LOCAL = "LOCAL",
  LOCALIZED = "LOCALIZED",
}

export type ReferenceWithRelationType<T> = T & {
  relationType: RelationType;
};

export type ImportIdentifiable = { importId: string };
