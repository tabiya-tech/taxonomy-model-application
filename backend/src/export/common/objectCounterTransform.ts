import { Transform, TransformCallback } from "stream";

export class ObjectCounterTransform extends Transform {
  private count: number = 0;

  constructor() {
    super({ objectMode: true });
  }

  _transform(chunk: unknown, _encoding: BufferEncoding, callback: TransformCallback): void {
    try {
      this.count++;
      this.push(chunk);
      callback(); // Always call callback after processing a chunk
    } catch (e: unknown) {
      // Proper error handling for streams
      callback(e as Error);
    }
  }

  getObjectCount(): number {
    return this.count;
  }

  _flush(callback: TransformCallback): void {
    callback();
  }
}
