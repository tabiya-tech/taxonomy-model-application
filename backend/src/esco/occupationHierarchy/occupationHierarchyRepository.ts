import mongoose from "mongoose";
import { ObjectTypes } from "esco/common/objectTypes";
import { IOccupationDoc } from "esco/occupations/occupation.types";
import { IOccupationGroupDoc } from "esco/occupationGroup/OccupationGroup.types";
import {
  INewOccupationHierarchyPairSpec,
  IOccupationHierarchyPair,
  IOccupationHierarchyPairDoc,
} from "./occupationHierarchy.types";

import { isNewOccupationHierarchyPairSpecValid, isParentChildCodeConsistent } from "./occupationHierarchyValidation";
import { getModelName } from "esco/common/mongooseModelNames";
import { handleInsertManyError } from "esco/common/handleInsertManyErrors";
import { Readable } from "node:stream";
import stream from "stream";
import { DocumentToObjectTransformer } from "esco/common/documentToObjectTransformer";
import { OccupationGroupModelPaths } from "esco/occupationGroup/OccupationGroupModel";
import { OccupationModelPaths } from "esco/occupations/occupationModel";

export interface IOccupationHierarchyRepository {
  readonly hierarchyModel: mongoose.Model<IOccupationHierarchyPairDoc>;
  readonly occupationGroupModel: mongoose.Model<IOccupationGroupDoc>;
  readonly occupationModel: mongoose.Model<IOccupationDoc>;

  /**
   * Creates multiple new OccupationHierarchyPair entries.
   *
   * @param {string} modelId - The modelId of the model to which the new OccupationHierarchyPair entries will belong.
   * @param {INewOccupationHierarchyPairSpec[]} newOccupationHierarchyPairSpecs - An array of specifications for the new OccupationHierarchyPair entries.
   * @return {Promise<IOccupationHierarchyPair[]>} - A Promise that resolves to an array containing the newly created IOccupationHierarchyPair entries.
   * Excludes entries that fail validation and returns a subset of successfully created entries.
   * Rejects with an error if any entry cannot be created due to reasons other than validation.
   */
  createMany(
    modelId: string,
    newOccupationHierarchyPairSpecs: INewOccupationHierarchyPairSpec[]
  ): Promise<IOccupationHierarchyPair[]>;

  /**
   * Returns all OccupationHierarchyPair entries as a stream. The entries are transformed to objects (via the .toObject()).
   * @param {string} modelId - The modelId of the occupations.
   * @return {Readable} - A Readable stream of IOccupationHierarchyPair
   * Rejects with an error if the operation fails.
   */
  findAll(modelId: string): Readable;
}

export class OccupationHierarchyRepository implements IOccupationHierarchyRepository {
  public readonly hierarchyModel: mongoose.Model<IOccupationHierarchyPairDoc>;
  public readonly occupationGroupModel: mongoose.Model<IOccupationGroupDoc>;
  public readonly occupationModel: mongoose.Model<IOccupationDoc>;

  constructor(
    hierarchyModel: mongoose.Model<IOccupationHierarchyPairDoc>,
    occupationGroupModel: mongoose.Model<IOccupationGroupDoc>,
    occupationModel: mongoose.Model<IOccupationDoc>
  ) {
    this.hierarchyModel = hierarchyModel;
    this.occupationGroupModel = occupationGroupModel;
    this.occupationModel = occupationModel;
  }

  async createMany(
    modelId: string,
    newOccupationHierarchyPairSpecs: INewOccupationHierarchyPairSpec[]
  ): Promise<IOccupationHierarchyPair[]> {
    if (!mongoose.Types.ObjectId.isValid(modelId)) throw new Error(`Invalid modelId: ${modelId}`);
    const newHierarchyDocs: mongoose.Document<unknown, unknown, IOccupationHierarchyPairDoc>[] = [];
    try {
      const existingIds = new Map<string, ObjectTypes[]>();
      // The idToCode map was added after the existing ids as a separate map to avoid changing
      // the isNewOccupationHierarchyPairSpecValid function
      const idToCode = new Map<string, { type: ObjectTypes; code: string }[]>();

      //  get all Occupation groups
      const _existingOccupationGroupIds = await this.occupationGroupModel
        .find({ modelId: { $eq: modelId } })
        .select(`_id ${OccupationGroupModelPaths.groupType} ${OccupationGroupModelPaths.code}`)
        .exec();
      _existingOccupationGroupIds.forEach((occupationGroup) => {
        existingIds.set(occupationGroup._id.toString(), [occupationGroup.groupType]);
        idToCode.set(occupationGroup._id.toString(), [{ type: occupationGroup.groupType, code: occupationGroup.code }]);
      });

      //  get all Occupations
      const _existingOccupations = await this.occupationModel
        .find({ modelId: { $eq: modelId } })
        .select(`_id ${OccupationModelPaths.occupationType} ${OccupationModelPaths.code}`)
        .exec();

      // beside the id, we also need to know the occupationType and the code
      _existingOccupations.forEach((occupation) => {
        const foundOccupationGroupIds = existingIds.get(occupation._id.toString());
        const foundCodes = idToCode.get(occupation._id.toString());
        if (foundOccupationGroupIds) {
          foundOccupationGroupIds.push(occupation.occupationType);
        }
        if (foundCodes) {
          foundCodes.push({ type: occupation.occupationType, code: occupation.code });
          return;
        }

        existingIds.set(occupation._id.toString(), [occupation.occupationType]);
        idToCode.set(occupation._id.toString(), [{ type: occupation.occupationType, code: occupation.code }]);
      });

      const newOccupationHierarchyPairModels = newOccupationHierarchyPairSpecs
        .filter((spec) => {
          const validType = isNewOccupationHierarchyPairSpecValid(spec, existingIds);
          if (!validType) {
            console.warn("OccupationHierarchyRepository.createMany: invalid entry", spec);
            return false;
          }
          const validCode = isParentChildCodeConsistent(
            spec.parentType,
            spec.parentId,
            spec.childType,
            spec.childId,
            idToCode
          );
          if (!validCode) {
            console.warn("OccupationHierarchyRepository.createMany: invalid entry", spec);
          }
          return validType && validCode;
        })
        .map((spec) => {
          try {
            return new this.hierarchyModel({
              ...spec,
              modelId: modelId,
              parentDocModel: getModelName(spec.parentType),
              childDocModel: getModelName(spec.childType),
            });
          } catch (e: unknown) {
            console.warn("OccupationHierarchyRepository.createMany: invalid entry", spec, e);
            return null;
          }
        })
        .filter(Boolean);

      const docs = await this.hierarchyModel.insertMany(newOccupationHierarchyPairModels, {
        ordered: false,
      });
      newHierarchyDocs.push(...docs);
    } catch (e: unknown) {
      const docs = handleInsertManyError<IOccupationHierarchyPairDoc>(
        e,
        "OccupationHierarchyRepository.createMany",
        newOccupationHierarchyPairSpecs.length
      );
      newHierarchyDocs.push(...docs);
    }

    if (newOccupationHierarchyPairSpecs.length !== newHierarchyDocs.length) {
      console.warn(
        `OccupationHierarchyRepository.createMany: ${
          newOccupationHierarchyPairSpecs.length - newHierarchyDocs.length
        } invalid entries were not created`
      );
    }
    return newHierarchyDocs.map((pair) => pair.toObject());
  }

  findAll(modelId: string): Readable {
    try {
      const pipeline = stream.pipeline(
        // use $eq to prevent NoSQL injection
        this.hierarchyModel.find({ modelId: { $eq: modelId } }).cursor(),
        new DocumentToObjectTransformer<IOccupationHierarchyPair>(),
        () => undefined
      );
      pipeline.on("error", (e) => {
        console.error(new Error("OccupationHierarchyRepository.findAll: stream failed", { cause: e }));
      });

      return pipeline;
    } catch (e: unknown) {
      const err = new Error("OccupationHierarchyRepository.findAll: findAll failed", { cause: e });
      console.error(err);
      throw err;
    }
  }
}
