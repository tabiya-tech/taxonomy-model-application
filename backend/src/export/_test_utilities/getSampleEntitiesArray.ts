import { INewOccupationSpec, IOccupation } from "esco/occupation/occupation.types";
import { getMockRandomISCOGroupCode } from "_test_utilities/mockISCOCode";
import { getMockRandomOccupationCode } from "_test_utilities/mockOccupationCode";
import { ObjectTypes, OccupationType, RelationType } from "esco/common/objectTypes";
import { INewLocalizedOccupationSpec } from "esco/localizedOccupation/localizedOccupation.types";
import { INewISCOGroupSpec } from "esco/iscoGroup/ISCOGroup.types";
import { INewSkillSpec, ISkill, ReuseLevel, SkillType } from "esco/skill/skills.types";
import { INewSkillGroupSpec, ISkillGroup } from "esco/skillGroup/skillGroup.types";
import { getMockRandomSkillCode } from "_test_utilities/mockSkillGroupCode";
import { getMockStringId } from "_test_utilities/mockMongoId";
import { INewOccupationHierarchyPairSpec } from "esco/occupationHierarchy/occupationHierarchy.types";
import { INewSkillHierarchyPairSpec } from "esco/skillHierarchy/skillHierarchy.types";
import { INewOccupationToSkillPairSpec } from "esco/occupationToSkillRelation/occupationToSkillRelation.types";
import { INewSkillToSkillPairSpec } from "esco/skillToSkillRelation/skillToSkillRelation.types";

export const getSampleOccupationSpecs = (givenModelId: string, isLocal: boolean = false): INewOccupationSpec[] => {
  return Array.from<never, INewOccupationSpec>({ length: 100 }, (_, i) => ({
    ISCOGroupCode: getMockRandomISCOGroupCode(),
    definition: `definition_${i}`,
    regulatedProfessionNote: `regulatedProfessionNote_${i}`,
    scopeNote: `scopeNote_${i}`,
    altLabels: i % 2 ? [`altLabel_1`, `altLabel_2`] : [],
    code: getMockRandomOccupationCode(isLocal),
    preferredLabel: `Occupation_${i}`,
    modelId: givenModelId,
    originUUID: ``,
    ESCOUri: `ESCOUri_${i}`,
    description: `description_${i}`,
    importId: `importId_${i}`,
    occupationType: isLocal ? OccupationType.LOCAL : OccupationType.ESCO,
  }));
};
export const getSampleLocalizedOccupationSpecs = (
  givenEscoOccupations: IOccupation[]
): INewLocalizedOccupationSpec[] => {
  return givenEscoOccupations.map((occupation, i) => ({
    localizesOccupationId: occupation.id,
    altLabels: i % 2 ? [] : [`altLabel_1`, `altLabel_2`],
    modelId: occupation.modelId,
    description: `description_${i}`,
    importId: `importId_${i}`,
    occupationType: OccupationType.LOCALIZED,
  }));
};

export const getSampleISCOGroupSpecs = (givenModelId: string): INewISCOGroupSpec[] => {
  return Array.from<never, INewISCOGroupSpec>({ length: 100 }, (_, i) => ({
    code: getMockRandomISCOGroupCode(),
    preferredLabel: `ISCOGroup_${i}`,
    modelId: givenModelId,
    originUUID: ``,
    altLabels: i % 2 ? [`altLabel_1`, `altLabel_2`] : [],
    ESCOUri: `ESCOUri_${i}`,
    description: `description_${i}`,
    importId: `importId_${i}`,
  }));
};

export const getSampleSkillsSpecs = (givenModelId: string): INewSkillSpec[] => {
  return Array.from<never, INewSkillSpec>({ length: 100 }, (_, i) => ({
    preferredLabel: `Skill_${i}`,
    modelId: givenModelId,
    originUUID: ``,
    ESCOUri: `ESCOUri_${i}`,
    definition: `definition_${i}`,
    description: `description_${i}`,
    scopeNote: `scopeNote_${i}`,
    skillType: SkillType.Knowledge,
    reuseLevel: ReuseLevel.CrossSector,
    altLabels: i % 2 ? [`altLabel_1`, `altLabel_2`] : [],
    importId: `importId_${i}`,
  }));
};

export const getSampleSkillGroupsSpecs = (givenModelId: string): INewSkillGroupSpec[] => {
  return Array.from<never, INewSkillGroupSpec>({ length: 100 }, (_, i) => ({
    code: getMockRandomSkillCode(),
    preferredLabel: `SkillGroup_${i}`,
    modelId: givenModelId,
    originUUID: ``,
    ESCOUri: `ESCOUri_${i}`,
    description: `description_${i}`,
    scopeNote: `scopeNote_${i}`,
    altLabels: i % 2 ? [`altLabel_1`, `altLabel_2`] : [],
    importId: `importId_${i}`,
  }));
};

export const getSampleOccupationHierarchy = (givenModelId: string): INewOccupationHierarchyPairSpec[] => {
  return Array.from<never, INewOccupationHierarchyPairSpec>({ length: 100 }, (_, i) => ({
    parentId: getMockStringId(999),
    parentType: ObjectTypes.Occupation,
    childType: ObjectTypes.Occupation,
    childId: getMockStringId(i),
    modelId: givenModelId,
    importId: `importId_${i}`,
  }));
};

export const getSampleSkillsHierarchy = (
  givenSkills: ISkill[],
  givenSkillGroups: ISkillGroup[]
): INewSkillHierarchyPairSpec[] => {
  return givenSkills.map((skill) => ({
    parentId: givenSkillGroups[0].id,
    parentType: ObjectTypes.SkillGroup,
    childType: ObjectTypes.Skill,
    childId: skill.id,
    modelId: skill.modelId,
  }));
};

export const getSampleOccupationToSkillRelations = (
  givenOccupations: IOccupation[],
  givenSkills: ISkill[]
): INewOccupationToSkillPairSpec[] => {
  return givenSkills.map((skill, i) => ({
    requiringOccupationType: OccupationType.LOCAL,
    requiredSkillId: skill.id,
    requiringOccupationId: givenOccupations[i].id,
    relationType: RelationType.OPTIONAL,
  }));
};

export const getSampleSkillToSkillRelations = (givenSkills: ISkill[]): INewSkillToSkillPairSpec[] => {
  return givenSkills.map((skill, i) => ({
    requiringSkillId: givenSkills[i + 1]?.id ?? givenSkills[0].id,
    requiredSkillId: skill.id,
    relationType: RelationType.OPTIONAL,
  }));
};
