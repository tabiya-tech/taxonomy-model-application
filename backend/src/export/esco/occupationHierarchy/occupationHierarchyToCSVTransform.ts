import { stringify } from "csv-stringify";
import { pipeline, Transform } from "stream";
import { IOccupationHierarchyExportRow, occupationHierarchyExportHeaders } from "esco/common/entityToCSV.types";
import { getRepositoryRegistry } from "server/repositoryRegistry/repositoryRegistry";
import { IOccupationHierarchyPair } from "esco/occupationHierarchy/occupationHierarchy.types";
import { Readable } from "node:stream";

export type IUnpopulatedOccupationHierarchy = Omit<IOccupationHierarchyPair, "parentDocModel" | "childDocModel">;

export const transformOccupationHierarchySpecToCSVRow = (
  occupationHierarchy: IUnpopulatedOccupationHierarchy
): IOccupationHierarchyExportRow => {
  return {
    PARENTOBJECTTYPE: occupationHierarchy.parentType,
    PARENTID: occupationHierarchy.parentId,
    CHILDID: occupationHierarchy.childId,
    CHILDOBJECTTYPE: occupationHierarchy.childType,
    CREATEDAT: occupationHierarchy.createdAt.toISOString(),
    UPDATEDAT: occupationHierarchy.updatedAt.toISOString(),
  };
};

class OccupationHierarchyToCSVRowTransformer extends Transform {
  constructor() {
    super({ objectMode: true });
  }

  _transform(
    skillHierarchy: IUnpopulatedOccupationHierarchy,
    _encoding: BufferEncoding,
    callback: (error?: Error | null, data?: never) => void
  ): void {
    try {
      const transformedRow = transformOccupationHierarchySpecToCSVRow(skillHierarchy);
      this.push(transformedRow);
      callback();
    } catch (e: unknown) {
      // Make sure stringification doesn't fail, otherwise throwing an error will cause the stream to hang
      let json: string = "";
      try {
        json = JSON.stringify(skillHierarchy, null, 2);
      } finally {
        const error = new Error(`Failed to transform OccupationHierarchy to CSV row: ${json}`, { cause: e });
        console.error(error);
        callback(error);
      }
    }
  }
}

const OccupationHierarchyToCSVTransform = (modelId: string): Readable => {
  // the stringify is a stream, and we need a new one every time we create a new pipeline
  const occupationHierarchyStringifier = stringify({
    header: true,
    columns: occupationHierarchyExportHeaders,
  });

  return pipeline(
    getRepositoryRegistry().occupationHierarchy.findAll(modelId),
    new OccupationHierarchyToCSVRowTransformer(),
    occupationHierarchyStringifier,
    (error) => {
      if (error) {
        console.error(new Error("Transforming OccupationHierarchy to CSV failed", { cause: error }));
      }
    }
  );
};

export default OccupationHierarchyToCSVTransform;
