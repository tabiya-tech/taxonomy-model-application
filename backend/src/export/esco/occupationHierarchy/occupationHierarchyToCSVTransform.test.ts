// Mute chatty console logs
import "_test_utilities/consoleMock";

import { getRepositoryRegistry } from "server/repositoryRegistry/repositoryRegistry";
import { getMockStringId } from "_test_utilities/mockMongoId";
import { ObjectTypes } from "esco/common/objectTypes";
import { Readable } from "node:stream";
import occupationHierarchyToCSVTransform, * as occupationHierarchyToCSVTransformModule from "./occupationHierarchyToCSVTransform";
import { IUnpopulatedOccupationHierarchy } from "./occupationHierarchyToCSVTransform";
import { IOccupationHierarchyRepository } from "esco/occupationHierarchy/occupationHierarchyRepository";
import { parse } from "csv-parse/sync";

const OccupationHierarchyRepositorySpy = jest.spyOn(getRepositoryRegistry(), "occupationHierarchy", "get");

const getMockOccupationHierarchies = (): IUnpopulatedOccupationHierarchy[] => {
  return Array.from({ length: 6 }, (_, i) => ({
    id: getMockStringId(i * 3),
    modelId: getMockStringId(1),
    parentId: getMockStringId(i * 3 + 1),
    childId: getMockStringId(i * 3 + 2),
    parentType: i % 2 ? ObjectTypes.Occupation : ObjectTypes.ISCOGroup,
    childType: i % 2 ? ObjectTypes.ISCOGroup : ObjectTypes.Occupation,
    createdAt: new Date(),
    updatedAt: new Date(),
  }));
};

function setupOccupationHierarchyRepositoryMock(findAllImpl: () => Readable) {
  const mockOccupationHierarchyRepository: IOccupationHierarchyRepository = {
    hierarchyModel: undefined as any,
    occupationModel: undefined as any,
    iscoGroupModel: undefined as any,
    createMany: jest.fn().mockResolvedValue(null),
    findAll: jest.fn().mockImplementationOnce(findAllImpl),
  };
  OccupationHierarchyRepositorySpy.mockReturnValue(mockOccupationHierarchyRepository);
}

describe("occupationHierarchyToCSVTransform", () => {
  test("should correctly transform occupationHierarchy data to CSV", async () => {
    // GIVEN findAll returns a stream of occupationHierarchies
    const givenHierarchies = getMockOccupationHierarchies();
    setupOccupationHierarchyRepositoryMock(() => Readable.from(givenHierarchies));

    // WHEN the transformation is applied
    const transformedStream = occupationHierarchyToCSVTransform("foo");

    // THEN expect the output to be a valid CSV
    const chunks = [];
    for await (const chunk of transformedStream) {
      chunks.push(chunk);
    }
    const actualCSVOutput = chunks.join("");
    const parsedObjects = parse(actualCSVOutput, { columns: true });
    // AND contain the given occupationHierarchy data
    expect(parsedObjects).toMatchSnapshot();
    expect(actualCSVOutput).toMatchSnapshot();
    // AND the stream should end
    expect(transformedStream.closed).toBe(true);
  });

  describe("handle errors during stream processing", () => {
    test("should log an error and end the stream when the source repository fails", async () => {
      // GIVEN that the source occupationHierarchy stream will emit an error
      const givenError = new Error("Test Error");
      setupOccupationHierarchyRepositoryMock(
        () =>
          new Readable({
            read() {
              this.emit("error", givenError);
            },
          })
      );

      // WHEN the transformation stream is consumed
      const transformedStream = occupationHierarchyToCSVTransform("foo");

      // THEN expect the given error to be thrown
      await expect(async () => {
        //  iterate to consume the stream
        for await (const _ of transformedStream) {
          // do nothing
        }
      }).rejects.toThrowError(givenError);
      // AND the error should be logged
      expect(console.error).toHaveBeenNthCalledWith(1, expect.any(Error), expect.any(Error));
      // AND the stream should end
      expect(transformedStream.closed).toBe(true);
    });

    test("should log an error and end the stream when the CSV stringifier fails", async () => {
      // GIVEN findAll returns a stream of occupationHierarchys
      setupOccupationHierarchyRepositoryMock(() => Readable.from(getMockOccupationHierarchies()));

      // AND  the transformOccupationHierarchySpecToCSVRow will throw an error
      const givenError = new Error("Mocked Transformation Error");
      jest
        .spyOn(occupationHierarchyToCSVTransformModule, "transformOccupationHierarchySpecToCSVRow")
        .mockImplementationOnce((_: occupationHierarchyToCSVTransformModule.IUnpopulatedOccupationHierarchy) => {
          throw givenError;
        });

      // WHEN the transformation stream is consumed
      const transformedStream = occupationHierarchyToCSVTransform("foo");

      // THEN expect the given error to be thrown
      await expect(async () => {
        //  iterate to consume the stream
        for await (const _ of transformedStream) {
          // do nothing
        }
      }).rejects.toThrowError("Failed to transform OccupationHierarchy to CSV row");
      // AND the error to be logged
      expect(console.error).toHaveBeenNthCalledWith(1, expect.any(Error), expect.any(Error));
      expect(console.error).toHaveBeenNthCalledWith(2, expect.any(Error), expect.any(Error));
      // AND the stream to end
      expect(transformedStream.closed).toBe(true);
    });
  });
});
