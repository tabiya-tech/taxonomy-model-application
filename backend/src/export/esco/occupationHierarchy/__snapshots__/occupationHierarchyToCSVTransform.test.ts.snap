// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`occupationHierarchyToCSVTransform should correctly transform occupationHierarchy data to CSV 1`] = `
[
  {
    "CHILDID": "000000000000000000000002",
    "CHILDOBJECTTYPE": "escooccupation",
    "CREATEDAT": "1970-01-01T00:00:00.000Z",
    "PARENTID": "000000000000000000000001",
    "PARENTOBJECTTYPE": "iscogroup",
    "UPDATEDAT": "1970-01-01T00:00:00.000Z",
  },
  {
    "CHILDID": "000000000000000000000005",
    "CHILDOBJECTTYPE": "localoccupation",
    "CREATEDAT": "1970-01-01T00:00:00.001Z",
    "PARENTID": "000000000000000000000004",
    "PARENTOBJECTTYPE": "escooccupation",
    "UPDATEDAT": "1970-01-01T00:00:00.001Z",
  },
  {
    "CHILDID": "000000000000000000000008",
    "CHILDOBJECTTYPE": "escooccupation",
    "CREATEDAT": "1970-01-01T00:00:00.002Z",
    "PARENTID": "000000000000000000000007",
    "PARENTOBJECTTYPE": "iscogroup",
    "UPDATEDAT": "1970-01-01T00:00:00.002Z",
  },
  {
    "CHILDID": "00000000000000000000000b",
    "CHILDOBJECTTYPE": "localoccupation",
    "CREATEDAT": "1970-01-01T00:00:00.003Z",
    "PARENTID": "00000000000000000000000a",
    "PARENTOBJECTTYPE": "escooccupation",
    "UPDATEDAT": "1970-01-01T00:00:00.003Z",
  },
  {
    "CHILDID": "00000000000000000000000e",
    "CHILDOBJECTTYPE": "escooccupation",
    "CREATEDAT": "1970-01-01T00:00:00.004Z",
    "PARENTID": "00000000000000000000000d",
    "PARENTOBJECTTYPE": "iscogroup",
    "UPDATEDAT": "1970-01-01T00:00:00.004Z",
  },
  {
    "CHILDID": "000000000000000000000011",
    "CHILDOBJECTTYPE": "localoccupation",
    "CREATEDAT": "1970-01-01T00:00:00.005Z",
    "PARENTID": "000000000000000000000010",
    "PARENTOBJECTTYPE": "escooccupation",
    "UPDATEDAT": "1970-01-01T00:00:00.005Z",
  },
]
`;

exports[`occupationHierarchyToCSVTransform should correctly transform occupationHierarchy data to CSV 2`] = `
"PARENTOBJECTTYPE,PARENTID,CHILDID,CHILDOBJECTTYPE,CREATEDAT,UPDATEDAT
iscogroup,000000000000000000000001,000000000000000000000002,escooccupation,1970-01-01T00:00:00.000Z,1970-01-01T00:00:00.000Z
escooccupation,000000000000000000000004,000000000000000000000005,localoccupation,1970-01-01T00:00:00.001Z,1970-01-01T00:00:00.001Z
iscogroup,000000000000000000000007,000000000000000000000008,escooccupation,1970-01-01T00:00:00.002Z,1970-01-01T00:00:00.002Z
escooccupation,00000000000000000000000a,00000000000000000000000b,localoccupation,1970-01-01T00:00:00.003Z,1970-01-01T00:00:00.003Z
iscogroup,00000000000000000000000d,00000000000000000000000e,escooccupation,1970-01-01T00:00:00.004Z,1970-01-01T00:00:00.004Z
escooccupation,000000000000000000000010,000000000000000000000011,localoccupation,1970-01-01T00:00:00.005Z,1970-01-01T00:00:00.005Z
"
`;

exports[`occupationHierarchyToCSVTransform test transformOccupationHierarchySpecToCSVRow() should transform a OccupationHierarchy to a CSV row 1`] = `
{
  "CHILDID": "000000000000000000000002",
  "CHILDOBJECTTYPE": "escooccupation",
  "CREATEDAT": "1970-01-01T00:00:00.000Z",
  "PARENTID": "000000000000000000000001",
  "PARENTOBJECTTYPE": "iscogroup",
  "UPDATEDAT": "1970-01-01T00:00:00.000Z",
}
`;
